pkg/knowledge_base/edge_kb.go

```diff
@@ -221,74 +226,72 @@ func (kb EdgeKB) ExpandEdges(dag *core.ResourceGraph, appName string) (err error
 
 		zap.S().Debugf("Found valid paths %s", validPaths)
 		var validPath []Edge
-
-		// If we have more than 1 valid path we will always default to the direct path if it exists, otherwise we will raise an error since we cannot determine which path we are supposed to use.
-		if len(validPaths) > 1 {
-			for _, p := range validPaths {
-				if len(p) == 1 {
-					zap.S().Debug("Defaulting to direct path")
-					validPath = p
-				}
-			}
+		// Get the shortest route that satisfied constraints
+		for _, path := range validPaths {
 			if len(validPath) == 0 {
-				merr.Append(fmt.Errorf("found multiple paths which satisfy constraints for edge %s -> %s. \n Paths: %s", dep.Source.Id(), dep.Destination.Id(), validPaths))
+				validPath = path
+			} else if len(path) < len(validPath) {
+				validPath = path
+			} else if len(path) == len(validPath) {
+				merr.Append(fmt.Errorf("found multiple paths which satisfy constraints for edge %s -> %s and are the same length. \n Paths: %s", dep.Source.Id(), dep.Destination.Id(), validPaths))
+				continue
 			}
-		} else if len(validPaths) == 0 {
+		}
+		if len(validPath) == 0 {
 			merr.Append(fmt.Errorf("found no paths which satisfy constraints for edge %s -> %s. \n Paths: %s", dep.Source.Id(), dep.Destination.Id(), validPaths))
-		} else {
-			validPath = validPaths[0]
+			continue
+		}
 
-			// If the valid path is not the original direct path, we want to remove the initial direct dependency so we can fill in the new edges with intermediate nodes
-			if len(validPath) > 1 {
-				zap.S().Debugf("Removing dependency from %s -> %s", dep.Source.Id(), dep.Destination.Id())
-				err := dag.RemoveDependency(dep.Source.Id(), dep.Destination.Id())
-				if err != nil {
-					merr.Append(err)
-					continue
-				}
+		// If the valid path is not the original direct path, we want to remove the initial direct dependency so we can fill in the new edges with intermediate nodes
+		if len(validPath) > 1 {
+			zap.S().Debugf("Removing dependency from %s -> %s", dep.Source.Id(), dep.Destination.Id())
```

Comment:
nit: I'm surprised this isn't already logged in `RemoveDependency` - do we log inside or outside, I forget

Reply:
theres no logging in our graph calls for this, do you prefer that i move it into there?